storage/reader.go

```diff
@@ -392,6 +574,72 @@ func (r *Reader) Read(p []byte) (int, error) {
 	return n, err
 }
 
+func (r *Reader) readWithGRPC(p []byte) (int, error) {
+	// No stream to read from, either never initiliazed or Close was called.
+	// TODO: should an error be returned here? "reader has been closed"?
+	if r.stream == nil {
+		return 0, io.EOF
+	}
+
+	// The entire object has been read by this reader, return EOF.
+	if r.size != 0 && r.size == r.seen {
+		return 0, io.EOF
+	}
+
+	n := copy(p, r.leftovers)
+	r.seen += int64(n)
+	r.leftovers = r.leftovers[n:]
+
+	for len(p[n:]) > 0 {
+		msg, err := r.stream.Recv()
+		if err != nil {
+			if err == io.EOF && r.seen == r.size {
+				// Let the next call to readWithGRPC return EOF.
+				return n, nil
+			}
+			// This prevents reopening of the stream when the error is not
+			// retryable.
+			if !shouldRetry(err) {
+				return n, err
+			}
+
+			// Save the original error to return instead of the reinit error.
+			e := err
+
+			// Close existing stream and initialize new stream with updated
+			// offset.
+			r.closeStream()
+
+			// This will immediately attempt to reopen the stream, but will
+			// backoff if further attempts are necessary.
+			var res *readStreamResponse
+			res, r.cancelStream, err = r.reopenWithGRPC(r.seen)
+			if err != nil {
+				// Cannot reinstantiate the stream so return the original error.
+				return n, e
+			}
+			r.stream = res.stream
+			// Reopened the stream and Recv'd the first message, so carry on.
+			msg = res.response
+		}
+
+		// TODO: Capture incremental CRC32C for this chunk.
+		content := msg.GetChecksummedData().GetContent()
+		cp := copy(p[n:], content)
+		leftover := len(content) - cp
+		if leftover > 0 {
+			// Wasn't able to copy all of the data in the message, store for
+			// future Read calls.
+			r.leftovers = make([]byte, leftover)
```

Comment:
Memory performance wise, do you think it would be problematic to be re-making this buffer (as a custom length) rather than re-using a buffer or having a pool? I don't think this is something we need to address now but maybe another TODO.

Reply:
Good point. I think we should follow up on it and I've added a TODO. The implementation is simple atm and just shrinks `leftovers` by making subslices as the data is copied over to the read buffer. But we could make use of a "leftovers cursor" instead, and resuse the buffer once we've copied it all, expanding it if necessary.

Reply:
Sounds good!